cmds/ueficat/ueficat.go

```diff
@@ -0,0 +1,109 @@
+// Copyright 2018 the LinuxBoot Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// ueficat will cat a set of files from a firmware volume.
+// The files are specified with a regular expression that is
+// matched against a file GUID.
+//
+// Synopsis:
+//     ueficat <romimage> pat [pat ...]
+//
+// Description:
+//     ueficat reads a firmware volume and, for every file in it, matches it
+//     against the set of patterns passed in the command line.
+//     The pat can be a simple re matching a GUID, or an re of the form re:re.
+//     The optional second re matches a name in one of the sections, typically
+//     an EFI_SECTION_USER_INTERFACE, though we don't currently check for section type.
+//     If the pat matches the guid and name, and the file has a section of type EFI_SECTION_RAW,
+//     that section is written to os.Stdout.
+//     For example, in one UEFI image we have, we can say
+//     ueficat 7:Initrd
+//     and the initrd
+//     is output to stdout. If we have a lot of confidence, we can even say .:Initrd or 7:I
+package main
+
+import (
+	"flag"
+	"io/ioutil"
+	"log"
+	"os"
+	"regexp"
+	"strings"
+
+	"github.com/linuxboot/fiano/pkg/uefi"
+	"github.com/linuxboot/fiano/pkg/visitors"
+)
+
+type pat struct {
+	guid *regexp.Regexp
+	name *regexp.Regexp
+}
+
+var pats []pat
+
+//   File  74696E69-6472-632E-7069-6F2F62696F73  EFI_FV_FILETYPE_FREEFORM           3354174
+//    Sec                                        EFI_SECTION_RAW                    3354116
+//    Sec  Initrd                                EFI_SECTION_USER_INTERFACE         18
+//    Sec                                        EFI_SECTION_VERSION                14
+func pred(f *uefi.File, name string) bool {
+	var guid = f.Header.UUID.String()
+	for _, pat := range pats {
+		if !pat.guid.MatchString(guid) {
+			continue
+		}
+		for _, s := range f.Sections {
```

Comment:
This won't traverse the encapsulated sections. See the comment below.

Reply:
I think the best bet the, in the pred function, is to find the sections, append them to a global, and ignore the Matches at the end of the program?